controller/api/public/http_server.go

```diff
@@ -24,10 +25,14 @@ var (
 	listServicesPath  = fullURLPathFor("ListServices")
 	tapByResourcePath = fullURLPathFor("TapByResource")
 	selfCheckPath     = fullURLPathFor("SelfCheck")
+	endpointsPath     = fullURLPathFor("Endpoints")
 )
 
 type handler struct {
-	grpcServer pb.ApiServer
+	// 1) public.Api
+	// 2) controller/discovery.Api
+	grpcServer      pb.ApiServer
+	discoveryServer discoveryPb.ApiServer
 }
```

Comment:
Rather than having this handler reference two separate grpc server implementations, my preference is to create an `APIServer` interface, to parallel the `APIClient` interface that you've added, and then update the existing `grpcServer` struct to serve that interface. That will allow us to define all public API endpoints in one place.

So I'd modify grpc_server.go as follows:

```diff
diff --git a/controller/api/public/grpc_server.go b/controller/api/public/grpc_server.go
index d069b43c..b088a1d7 100644
--- a/controller/api/public/grpc_server.go
+++ b/controller/api/public/grpc_server.go
@@ -10,6 +10,7 @@ import (
 	"github.com/golang/protobuf/ptypes/duration"
 	"github.com/linkerd/linkerd2/controller/api/util"
 	healthcheckPb "github.com/linkerd/linkerd2/controller/gen/common/healthcheck"
+	"github.com/linkerd/linkerd2/controller/gen/controller/discovery"
 	tapPb "github.com/linkerd/linkerd2/controller/gen/controller/tap"
 	pb "github.com/linkerd/linkerd2/controller/gen/public"
 	"github.com/linkerd/linkerd2/controller/k8s"
@@ -28,6 +29,7 @@ type (
 	grpcServer struct {
 		prometheusAPI       promv1.API
 		tapClient           tapPb.TapClient
+		discoveryClient     discovery.DiscoveryClient
 		k8sAPI              *k8s.API
 		controllerNamespace string
 		ignoredNamespaces   []string
@@ -48,18 +50,25 @@ const (
 	promClientCheckDescription = "control plane can talk to Prometheus"
 )
 
+type APIServer interface {
+	pb.ApiServer
+	discovery.DiscoveryServer
+}
+
 func newGrpcServer(
 	promAPI promv1.API,
 	tapClient tapPb.TapClient,
+	discoveryClient discovery.DiscoveryClient,
 	k8sAPI *k8s.API,
 	controllerNamespace string,
 	ignoredNamespaces []string,
 	singleNamespace bool,
-) *grpcServer {
+) APIServer {
 
 	grpcServer := &grpcServer{
 		prometheusAPI:       promAPI,
 		tapClient:           tapClient,
+		discoveryClient:     discoveryClient,
 		k8sAPI:              k8sAPI,
 		controllerNamespace: controllerNamespace,
 		ignoredNamespaces:   ignoredNamespaces,
@@ -268,3 +277,15 @@ func (s *grpcServer) ListServices(ctx context.Context, req *pb.ListServicesReque
 
 	return &pb.ListServicesResponse{Services: svcs}, nil
 }
+
+func (s *grpcServer) Endpoints(ctx context.Context, params *discovery.EndpointsParams) (*discovery.EndpointsResponse, error) {
+	log.Debugf("Endpoints request %+v", params)
+
+	rsp, err := s.discoveryClient.Endpoints(ctx, params)
+	if err != nil {
+		log.Errorf("endpoints request to proxy API failed: %s", err)
+		return nil, err
+	}
+
+	return rsp, nil
+}
```

(note that this diff reflects the "Api" => "Discovery" protobuf change that I suggested)

And then I'd change this file as follows:

```diff
diff --git a/controller/api/public/http_server.go b/controller/api/public/http_server.go
index 0f96d104..6eb7ff9f 100644
--- a/controller/api/public/http_server.go
+++ b/controller/api/public/http_server.go
@@ -29,10 +29,7 @@ var (
 )
 
 type handler struct {
-	// 1) public.Api
-	// 2) controller/discovery.Api
-	grpcServer      pb.ApiServer
-	discoveryServer discoveryPb.ApiServer
+	grpcServer APIServer
 }
 
 func (h *handler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
@@ -247,7 +244,7 @@ func fullURLPathFor(method string) string {
 }
 
 func (h *handler) handleEndpoints(w http.ResponseWriter, req *http.Request) {
-	rsp, err := h.discoveryServer.Endpoints(req.Context(), &discoveryPb.EndpointsParams{})
+	rsp, err := h.grpcServer.Endpoints(req.Context(), &discoveryPb.EndpointsParams{})
 	if err != nil {
 		writeErrorToHTTPResponse(w, err)
 		return
@@ -264,7 +261,7 @@ func NewServer(
 	addr string,
 	prometheusClient promApi.Client,
 	tapClient tapPb.TapClient,
-	discoveryClient discoveryPb.ApiClient,
+	discoveryClient discoveryPb.DiscoveryClient,
 	k8sAPI *k8s.API,
 	controllerNamespace string,
 	ignoredNamespaces []string,
@@ -274,12 +271,12 @@ func NewServer(
 		grpcServer: newGrpcServer(
 			promv1.NewAPI(prometheusClient),
 			tapClient,
+			discoveryClient,
 			k8sAPI,
 			controllerNamespace,
 			ignoredNamespaces,
 			singleNamespace,
 		),
-		discoveryServer: newDiscoveryServer(discoveryClient),
 	}
 
 	instrumentedHandler := prometheus.WithTelemetry(baseHandler)
```

That will allow you to remove the `controller/api/public/discovery_server.go` and controller/api/public/discovery_server_test.go` files that you're adding.